### PR TITLE
Inherit lifetimes for async fn instead of duplicating them.

### LINKED ISSUE #61949: Inherent async fn returning `Self` treats type's lifetime parameters as `'static`
`rustc 1.37.0-nightly (2887008e0 2019-06-12)` and playground's `2019-06-17 b25ee644971a168287ee`

[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=50cf396172d221ccf89b61e7c4719fbd)

```rust
#![feature(async_await)]

pub struct Foo<'a> {
    pub bar: &'a i32,
}

impl<'a> Foo<'a> {
    pub async fn new(bar: &'a i32) -> Self {
        Foo {
            bar
        }
    }
}
```

```
error[E0495]: cannot infer an appropriate lifetime due to conflicting requirements
 --> src/lib.rs:8:22
  |
8 |     pub async fn new(bar: &'a i32) -> Self {
  |                      ^^^
  |
note: first, the lifetime cannot outlive the lifetime 'a as defined on the impl at 7:6...
 --> src/lib.rs:7:6
  |
7 | impl<'a> Foo<'a> {
  |      ^^
  = note: ...so that the expression is assignable:
          expected &i32
             found &'a i32
  = note: but, the lifetime must be valid for the static lifetime...
  = note: ...so that the types are compatible:
          expected Foo<'_>
             found Foo<'static>
```

---

It works to either change the signature to take `bar: &'static i32`, or to change the body of the fn to use a static borrow like `bar: &5`. So the compiler really does want the function to return a `Foo<'static>`, even though `Self` is a `Foo<'a>`

---

The workaround is to not use `Self`:

```diff
-    pub async fn new(bar: &'a i32) -> Self {
+    pub async fn new(bar: &'a i32) -> Foo<'a> {
```

---

Mentoring notes: See [notes here on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/187312-wg-async-foundations/topic/.60Self.60.20and.20.60'static.60.20.2361949/near/170476808).

### PR DESCRIPTION
**Pull request by cjgillot**
The current desugaring of `async fn foo<'a>(&usize) -> &u8` is equivalent to
```rust
fn foo<'a, '0>(&'0 usize) -> foo<'static, 'static>::Opaque<'a, '0, '_>;
type foo<'_a, '_0>::Opaque<'a, '0, '1> = impl Future<Output = &'1 u8>;
```
following the RPIT model.

Duplicating all the inherited lifetime parameters and setting the inherited version to `'static` makes lowering more complex and causes issues like #61949. This PR removes the duplication of inherited lifetimes to directly use
```rust
fn foo<'a, '0>(&'0 usize) -> foo<'a, '0>::Opaque<'_>;
type foo<'a, '0>::Opaque<'1> = impl Future<Output = &'1 u8>;
```
following the TAIT model.

Fixes https://github.com/rust-lang/rust/issues/61949

### CODE REVIEW

**Review by spastorino — APPROVED** [review]
This looks good, there are some good cleanups. There rust-analyzer tests failing. On the other hand there are some things being done that are going in the opposite direction. In particular, not inheriting things at all.

**Comment by oli-obk** [comment]
@bors r-

r=me with CI happy

**Comment by cjgillot** [comment]
@oli-obk The last commit blesses the nll tests. There is a regression in the diagnostic quality, but I don't know where it's coming from. Do you have suggestions?

**Comment by cjgillot** [comment]
Latest push contains a commit from #92467.

File: src/test/ui/async-await/issue-75785-confusing-named-region.stderr [hunk 1]
```diff
@@ -1,14 +1,10 @@
 error[E0506]: cannot assign to `*x` because it is borrowed
   --> $DIR/issue-75785-confusing-named-region.rs:9:5
    |
-LL | pub async fn async_fn(x: &mut i32) -> (&i32, &i32) {
```
Comment (Aaron1011) [inline src/test/ui/async-await/issue-75785-confusing-named-region.stderr:4]: This seems like a regression - the new error message doesn't explain why the assignment is really a problem.

Comment (cjgillot) [reply src/test/ui/async-await/issue-75785-confusing-named-region.stderr:4]: I'm not sure I agree with you. The problem is that `x` is assigned to while being immutably borrowed into `y` which is used later. It does not have anything to do with the fact that `x` or `y` appear in the signature, does it?

Comment (WaffleLapkin) [reply src/test/ui/async-await/issue-75785-confusing-named-region.stderr:4]: This diff also removed the part where "later use" is shown through

**Comment by nikomatsakis** [comment]
Hmm, I feel like this may be going in the wrong direction. What I don't like is that this is complicating the "internal Rust" a bit by introducing a construct that doesn't otherwise exist (an opaque type that inherits generics from its parent). That said, as compared to RPIT (and RPITIT), I have fewer concerns because we actually *do* want to capture all the generics and predicates.

**Comment by oli-obk** [comment]
> What I don't like is that this is complicating the "internal Rust" a bit by introducing a construct that doesn't otherwise exist (an opaque type that inherits generics from its parent).

@spastorino expressed concern to me privately about this a few weeks back and I completely misunderstood... 

I completely overlooked that this PR was not changing https://github.com/rust-lang/rust/blob/5435b84df355ee73f04a0e51e6b6c8cff62df6ec/compiler/rustc_typeck/src/collect.rs#L1596 to get the actual parent, but was still pointing at the function. I thought the PR title was just a bit confusing, but the PR was doing the right thing. I don't think that anymore, but the diff to where I think we should be isn't that large anymore.

So basically, copying the example from the main post:

The current desugaring of `async fn foo<'a>(&usize) -> &u8` is equivalent to
```rust
fn foo<'a, '0>(&'0 usize) -> foo::bar<'static, 'static, 'a, '0, '_> {
    type bar<'_a, '_0, 'a, '0, '1> = impl Future<Output = &'1 u8>;
}
```

and the new desugaring should be

```rust
fn foo<'a, '0>(&'0 usize) -> foo::bar<'a, '0, '_> {
    type bar<'a, '0, '1> = impl Future<Output = &'1 u8>;
}
```
following the TAIT model that you could write yourself by hand if you could specify return types that are only defined within the function.

**Comment by cjgillot** [comment]
> Hmm, I feel like this may be going in the wrong direction. What I don't like is that this is complicating the "internal Rust" a bit by introducing a construct that doesn't otherwise exist (an opaque type that inherits generics from its parent). That said, as compared to RPIT (and RPITIT), I have fewer concerns because we actually do want to capture all the generics and predicates.

Opaque type already inherit generics from their parent.  RPIT and async inherit type and const generics.  Lifetimes have just been handled specially because (1) RPIT inherit only a subset of lifetimes, (2) async was modelled after RPIT.

My motivation in this PR is to avoid having to keep a list of in scope lifetimes during lowering.  This list is used for two things: recognizing in-band lifetimes, and async fns.  In-band lifetimes can be handled on the AST during late resolution, and are actually quite easy to implement in AST-based lifetime resolution.  Duplicating lifetimes for async fns is much harder.

Taking a step back, I think the reason we are having this discussion is because of HIR does not have the right abstraction level. Current HIR gives a good description of code bodies, but is ill-suited for everything that needs to coordinate multiple items (generics, signature checking, impl trait...).

I feel that all the work to extend impl-trait (RPITIT, TAITIT...) will eventually hit this wall. I suggest we tackle the design in two steps:
- How do we describe the outside of an item-like?
  This is currently done both in HIR and with the `generics_of`, `type_of`... queries. There is a hiatus between what HIR thinks, and what typeck thinks (e.g. this whole generics inheritance issue).
- How do we describe the inside of an item-like?
  There, the HIR desugared expression language is actually a good fit.

In this design question, I'll defend the view that we should stop describing items, generics and types in HIR, and that HIR should only describe function bodies and static/const expressions.

In the mean time, marking this PR as blocked.

**Comment by oli-obk** [comment]
> Opaque type already inherit generics from their parent.

Side note: they inherit them from the type alias that they are in, not from the surrounding item.

**Comment by nikomatsakis** [comment]
@cjgillot I think I agree that it would be better to have HIR focus on describing the bodies of functions and not so much the module-and-item structure; I'm not entirely sure what the implications are of that, though. It seems like there is some useful code for lowering at least that is still needed, but I could definitely see that we might implement the queries that describe the overall structure to read the AST and lower directly, sharing code.

(e.g., we need to be able to lower predicates and things)

**Comment by nikomatsakis** [comment]
I'm a bit confused @oli-obk by what problem you are aiming to solve there.

**Comment by oli-obk** [comment]
Sorry, I forgot my reasoning. 
The discussion about splitting ast_lowering into a body and an external part seems not like something that we can usefully hash out in the body of this PR, we should probably move it to zulip + a hackmd tracking the design we want.

My comment was more on a way to move forward here. I want to avoid lowering where bounds twice. Instead of going from

```rust
fn foo<T: Iterator>() -> impl PartialEq<<T as Iterator>::Item> { ... }
```

to

```rust
type Foo<T: Iterator> = impl PartialEq<<T as Iterator>::Item>;
fn foo<T: Iterator>() -> Foo<T> { ... }
```

we go to


```rust
type Foo<T> = impl PartialEq<<T as Iterator>::Item>;
fn foo<T: Iterator>() -> Foo<T> { ... }
```

(note the missing `Iterator` bound on `T`). Afaict that was the part that made lowering twice complex. Though, if we throw out an MCP for splitting lowering so that body lowering can refer to results of signature lowering and T-compiler agrees, then we can merge this PR as it is, as it does make some things a bit cleaner (notably eradicating the `'static` hack)

**Comment by oli-obk** [comment]
@bors r+

No matter what final design we come up with, as mentioned previously, it reduces complexity

**Comment by pnkfelix** [comment]
> Finished benchmarking commit ([3cfa4de](https://github.com/rust-lang/rust/commit/3cfa4def7c87d571bd46d92fed608edf8fad236e)): [comparison url](https://perf.rust-lang.org/compare.html?start=5d8767cb229b097fedb1dd4bd9420d463c37774f&end=3cfa4def7c87d571bd46d92fed608edf8fad236e).
> 
> **Summary**: This benchmark run shows 2 relevant regressions 😿 to instruction counts.
> 
>     * Average relevant regression: 1.7%
> 
>     * Largest regression in instruction counts: 2.0% on `full` builds of `issue-88862 check`
> 
> 
> If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).
> 
> **Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.
> 
> @rustbot label: +perf-regression

It looks to me like the regression identified by the rust-timer run from 4 days ago (i.e. when this landed) is exactly the same as the regression identified back when the PR was [first filed](https://github.com/rust-lang/rust/pull/91403#issuecomment-984171670).

But, I do not see any discussion of that performance regression. 😞 

Part of the reason to do the timer runs is to force PR *authors and reviewers* to acknowledge the fallout of the changes they are making, and to answer questions like:
 * Was this performance regression anticipated?
 * Is it justified based on the benefit of the PR?
 * Is there a plan (or even vague ideas) to address the regression, or is it expected to be a cost we will just pay?

**Comment by tmandry** [comment]
Status: Reverted in #94088
